### PR TITLE
Refactor Modulation Extension Namespace

### PR DESCRIPTION
**Pull request by Atheatos**
Refactor the modulation extension based on discussions from https://github.com/gnuradio/SigMF/pull/87:
 - rename the namespace to `signal` from `modulation`
 - add a free-form `label` field for a short-hand description of the signal under `signal:emission`
 - move the signal descriptor fields to an object `Detail`, nested under `signal:emission`
 - add `channel` to `Detail` so that the communication system's bandwidth channel can be notated
 - rename `system` to `standard` under `Detail`
 - add an object `Emitter`, nested under `signal:emission`, to describe the hardware driving the communication system
 - update examples

I also added a `class_variant` field to `Detail` to cover less straightforward modulations such as pi/4-DQPSK and GMSK. This wasn't discussed in the other PR or previous issues, but something like it is necessary for the signals that I have. I'm not sure it's the best approach since its contents are open to interpretation, however. For example, should its value be `G` or `gaussian` for GMSK? How about `pi/4` vs `π/4`? Maybe a separate variant field for each class, perhaps in a completely different extension?

Here's how it looks for pi/4-DQPSK in the example I added in the spec:
```json
"detail": {
    "type": "digital",
    "class": "psk",
    "order": 4,
    "symbol_variant": "differential",
    "class_variant": "pi/4"
}
```

### CODE REVIEW

**Comment by Teque5** [comment]
Nice Job. This looks pretty well thought-out.

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [inline extensions/signal.sigmf-ext.md:55]: I know this field seems relatively straight-forward, but it would be good to specifically call out it's meant to be user-defined shortform. Maybe something like:

```txt
A high-level shortform description useful to human readers (e.g., "wifi", "LTE")
```

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [inline extensions/signal.sigmf-ext.md:66]: We should probably follow the NTIA's example, here, and specify the units as "decimal degrees" for both long & lat. Near the top of @MichaelCotton's comment, [here](https://github.com/gnuradio/SigMF/pull/87#issuecomment-508749532)

**Review by bhilburn — COMMENTED** [review]
Excellent job, @Atheatos, taking all of the discussion and ideas over nearly 3 years of PR hacking and making a useful PR out of it! Thanks so much!

I think with a couple of minor fixes (see comments), this is ready for merge 👍

**Comment by Atheatos** [comment]
Happy to hear that it looks good; thanks for the reviews.

I don't have anything to add to your description of the `label` field, so I just added it verbatim. Since adding a "units" column wouldn't really make sense for that table, I just added it to the description.

**Comment by bhilburn** [comment]
Would like to get review from @jacobagilbert / @n-west, then want to merge ASAP.

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:55]: I believe the `label` field is extremely common and should actually be moved to core (we could also call this `description` to be consistent with `global`), but this is really something I think should be up a level and more generic.

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:56]: I think `signal:emitter` should be directly within the annotations object.

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:57]: Same as above, `signal:emission_detail` or `signal:emission` should be directly within the annotations object. This also removes one of the hierarchical levels making this a bit simpler.

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:66]: Not to delay merging, but I think we should come to a consensus on how we handle geospatial metadata before we merge this. If we are going to use GeoJSON we should just put a `location` field here and allow the same format.

If we end up sticking with discrete coordinate fields, i suggest we add `altitude` and define the format.

See #73 and #84

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:106]: This is minor, but VSB can also used with digital modulations (see: ATSC)

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:188]: Can we change this name to be more specific then? `channel_bw` or `channel_bandwidth`?

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [inline extensions/signal.sigmf-ext.md:198]: Awesome providing examples!!

**Review by jacobagilbert — CHANGES_REQUESTED** [review]
I really like this overall, very well thought out.

I do think that there is some overlap between this and some other objects and that we might be able to flatten this by one level before merging; comments are inline.

**Comment by Teque5** [comment]
Suggestion for extensions in general: consider section numbering similar to [how NTIA has adopted sigmf extensions](https://github.com/NTIA/sigmf-ns-ntia/blob/master/ntia-emitter.sigmf-ext.md).

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [reply extensions/signal.sigmf-ext.md:55]: This is an interesting point, @jacobagilbert. Currently, that's meant to be filled by the `comment` field, but I think what you're really highlighting is that it would be worthwhile to have a short-form human **and machine** -readable field that's in `core`. 

For purposes of process, I recommend we leave this field here, and then file a new issue & PR discussing and making that change across the codebase to keep the change atomic.

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [reply extensions/signal.sigmf-ext.md:66]: Agreed, this should be made consistent with the outcomes of that discussion. As with the `label` discussion above, I'd like to /not/ hold this PR on this topic, and then once that decision is made, make that change across the code base.

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [reply extensions/signal.sigmf-ext.md:106]: Interesting, I didn't know that! @Atheatos, want to make this `(analog / digital)` for completeness?

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [reply extensions/signal.sigmf-ext.md:188]: Good point, and agreed with @jacobagilbert.

File: extensions/signal.sigmf-ext.md
Comment (bhilburn) [reply extensions/signal.sigmf-ext.md:57]: Pulling from the examples section, this is what we have currently:

```json
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "signal:emission": {
            "label": "wifi",
            "detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
            },
            "emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
            }
        }
    }]
}
```

What you are suggesting is change it to what would look like:

```json
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "signal:emission": {
            "label": "wifi"
        },
        "signal:detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
         },
         "signal:emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
        }
    }]
}
```

I agree with you that it simplifies the "hierarchy" of reaching into the `emission`, but it also bloats what sits at the level of the hierarchy for the annotation, and from a semantic perspective, I actually think it makes less sense?

Not sold one way or the other. Thoughts from @n-west, @Teque5, @Atheatos?

**Review by bhilburn — CHANGES_REQUESTED** [review]
Responded to review comments from @jacobagilbert. I think there are two small changes to make, one outstanding discussion to close (see hierarchy of `emission`), and two other items that will be addressed outside of this PR.

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [reply extensions/signal.sigmf-ext.md:57]: This is what I am specifically proposing:

```JSON
{
    [...]
    "annotations": [{
        "core:sample_start": 0,
        "core:sample_count": 1000000,
        "core:label": "wifi",
        "signal:detail": {
                "type": "digital",
                "standard": "802.11ac",
                "channel": 8
         },
         "signal:emitter": {
                "seid": 1,
                "manufacturer": "linksys",
                "latitude": 0.000,
                "longitude": 0.000
        }
    }]
}

largely because in my use of SigMF annotations there are many times i would *only* have `label` or `signal:emitter` or `signal:detail` information to include and the extra layer of hierarchy. If we move `label` to `core` as i think we agree is a good idea, this is two functionally separate concepts (details about the emitter, and details about the emission).

File: extensions/signal.sigmf-ext.md
Comment (jacobagilbert) [reply extensions/signal.sigmf-ext.md:55]: Yes, thats right. 

Are you suggesting that we merge this here then separately move it to core after this is merged?

**Comment by jacobagilbert** [comment]
@Atheatos Do you have time to work on this? If not, are you OK if I hijack this to get the changes made for inclusion in v1.0?

**Comment by bhilburn** [comment]
@jacobagilbert - I think you're clear to pick this up and run with it. Please do it on top of @Atheatos's commits so his contributions remain represented in-tree.

**Comment by Atheatos** [comment]
@jacobagilbert @bhilburn 
Hey, sorry for the delay in responding. Indeed, I'm not able to work on this right now, so I appreciate you moving forward to get the change integrated for the v.1.0 release

**Comment by jacobagilbert** [comment]
Thanks @Atheatos I just opened a new PR though. It has your commit history so your contribution will be retained.

**Pull request closed without merging**
